plan/predicate_push_down.go

```diff
@@ -47,10 +48,281 @@ func columnSubstitute(expr expression.Expression, schema expression.Schema, newE
+	// we propagate constant as the following step:
+	// first: "1 = b and b = c and c = 1 and a = 1";
+	// next:  "1 = b and 1 = c and c = 1 and a = 1";
+	// next:  "1 = b and 1 = c and 1 = 1 and a = 1";
+	// next:  "1 = b and 1 = c and a = 1";
+
+	// e.g for condition: "a = b and b = c and b = 2 and a = 1";
+	// we propagate constant as the following step:
+	// first: "a = 2 and 2 = c and b = 2 and a = 1";
+	// next:  "a = 2 and 2 = c and b = 2 and 2 = 1";
+	// next:  "0"
+	isHandled := make([]bool, len(conditions))
+	equalities := make(map[string]map[int]*expression.Constant, 0) // transitive equality predicates between one column and one constant
+	for i := 0; i < len(conditions); i++ {
+		if isHandled[i] {
+			continue
+		}
+		if len(equalities) != 0 {
+			conditions[i] = constantSubstitute(equalities, conditions[i])
+		}
+		expr, ok := conditions[i].(*expression.ScalarFunction)
+		if !ok {
+			continue
+		}
+		// process the included OR conditions recursively to do the same for CNF item.
+		if expr.FuncName.L == ast.OrOr {
+			expressions := splitDNFItems(conditions[i])
+			newExpression := make([]expression.Expression, 0)
+			for _, v := range expressions {
+				newExpression = append(newExpression, propagateConstant([]expression.Expression{v})...)
+			}
+			conditions[i] = composeDNFConditions(newExpression)
+			isHandled[i] = true
+		} else if expr.FuncName.L == ast.AndAnd {
+			newExpression := propagateConstant(splitCNFItems(conditions[i]))
+			conditions[i] = expression.ComposeCNFCondition(newExpression)
+			isHandled[i] = true
+		} else if expr.FuncName.L == ast.EQ {
+			var (
+				col      *expression.Column
+				fromID   string
+				position int
+				val      *expression.Constant
+				isColumn bool
+			)
+			left, ok1 := expr.Args[0].(*expression.Constant)
+			right, ok2 := expr.Args[1].(*expression.Constant)
+			if col, isColumn = expr.Args[0].(*expression.Column); ok2 && isColumn {
+				val = right
+			} else if col, isColumn = expr.Args[1].(*expression.Column); ok1 && isColumn {
+				val = left
+			} else {
+				continue
+			}
+			fromID = col.FromID
+			position = col.Position
+			if _, ok = equalities[fromID]; !ok {
+				equalities[fromID] = make(map[int]*expression.Constant, 1)
+			}
+			equalities[fromID][position] = val
+			isHandled[i] = true
+			i = -1
+		}
+	}
+	for i := 0; i < len(conditions); i++ { // remove useless predicates, such as "1 and a = 1" ==> "a=1".
+		if v, ok := conditions[i].(*expression.Constant); ok {
+			if v.Value.IsNull() {
+				conditions = []expression.Expression{v}
+				return conditions
+			}
+			intVal, _ := v.Value.ToInt64()
+			if intVal < 1 && intVal > -1 {
+				conditions = []expression.Expression{&expression.Constant{Value: types.NewIntDatum(0)}}
+				return conditions
+			} else {
+				conditions = append(conditions[:i], conditions[i+1:]...)
+				i--
+			}
+		}
+	}
+	if len(conditions) == 0 {
+		conditions = []expression.Expression{&expression.Constant{Value: types.NewIntDatum(1)}}
```

Comment:
For expression, like expression  "2 = 2 and 1 = 1", I remove all of them in previous code, but now that these code is proved to be useless, there is no need to do this, too.